### PR TITLE
File/environment-based secrets

### LINKED ISSUE #31: Ability to provide Cloudflare API token as a Docker secret
**Is your feature request related to a problem? Please describe.**
At the moment you have to provide your API token within the config file, meaning you can't commit that into any source control system.

**Describe the solution you'd like**
The ability to provide the token directly from Docker secrets. A good example in the wild is how Authelia handles this: https://www.authelia.com/docs/configuration/secrets.html. For every sensitive config key there is a corresponding environment variable that tells the service which file to read a value from.

**Describe alternatives you've considered**
You could treat the whole config file as a Docker secret, but that doesn't solve the source control problem. I also thought about a janky script to edit the config when the container starts up and insert the token, but that would be unstable and would only benefit me rather than all users.

**Additional context**
Happy to work on a PR for this if it would be a welcome contribution.

### PR DESCRIPTION
**Pull request by markormesher**
Fixes #31

First draft of a way to provide some config values from a file or environment variable rather than hardcoding them in `config.json` that is fully backwards compatible. It works by replacing the constant values with nested objects as follows:

```json
"authentication": {
  "api_token": "my_secret_value"
}

// becomes...

"authentication": {
  "api_token": {
    "file": "/path/to/my/secret/file"
  }
}

// or...

"authentication": {
  "api_token": {
    "env": "MY_SECRET_ENV_VAR_NAME"
  }
}
```

I went with replacing the values rather than adding new keys like `api_token_file` because a) it lead to cleaner code and b) it avoids the problem of "what if a user enters both?".

Right now this is implemented for the different auth configs and the zone ID - I think those make the most sense but happy to take feedback. I've added some brief docs in the readme but I could also add example docker-compose files or similar.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -12,9 +12,9 @@ This script was written for the Raspberry Pi platform to enable low cost, simple
...
-| [![cloudflare-ddns docker image size](https://img.shields.io/docker/image-size/timothyjmiller/cloudflare-ddns?style=flat-square)](https://hub.docker.com/r/timothyjmiller/cloudflare-ddns "cloudflare-ddns docker image size")  | [![Total DockerHub pulls](https://img.shields.io/docker/pulls/timothyjmiller/cloudflare-ddns?style=flat-square)](https://hub.docker.com/r/timothyjmiller/cloudflare-ddns "Total DockerHub pulls")  | [![Official Discord Server](https://img.shields.io/discord/785778163887112192?style=flat-square)](https://discord.gg/UgGmwMvNxm "Official Discord Server")
+| Size                                                                                                                                                                                                                           | Downloads                                                                                                                                                                                         | Discord                                                                                                                                                    |
```
Comment (markormesher) [inline README.md:15]: Apologies, this is vim trying to help with MD auto-format on save. Happy to revert this if you want.

**Comment by timothymiller** [comment]
Thank you for putting this together. I have some feedback for some changes I'd like to see in this pr.

I do not want to see changes to the current `config.json` format. For supporting environment variables, we should add documentation for the various field names. 

Using the `config.json` approach, today you can run multiple zones + multiple subdomains within each zone with a single container. I want to keep this feature parity with the environment variables approach. We will have to use a numbered naming schema such as _2, _3, _4, etc probably up to 10. These environment variables will be loaded into a python object with the same format as `config.json`

That way, the user can write a single docker-compose file and manage all their DNS by way of environment variables.

File: README.md (hunk 1 above)
Comment (timothymiller) [reply README.md:15]: No problem, please fix it when you make your revisions.

**Comment by markormesher** [comment]
Was this meant to be closed? I've not had time to look at your suggestions yet but I would still like to work on this

**Comment by timothymiller** [comment]
Absolutely not. I am looking forward to your forthcoming commit.

I'll restate this again, because [we have multiple PRs trying to add file/environment based secrets](https://github.com/timothymiller/cloudflare-ddns/pull/41/commits/527590031a42b41ad396ec670244ce13b8713090), and so far they have came short of the community expectations.

A PR to support file/environment based secrets cannot break backwards compatibility with passing in a config.json file, which is the preferred method for Docker-compose & the legacy Python script method.

If we have an environment variable set, then we should short circuit and not read in the config.json file.

Is there anything else I can help explain to help you?

**Comment by markormesher** [comment]
Hey @timothymiller - sorry for the delay here, finally getting time to sit down and properly read through your comments.

100% with you on not breaking any backwards compatibility. The PR as it stands now doesn't do that because it only _extends_ the `config.json` format, so any previously-valid config would still work against this version. That includes configs for multiple domains with multiple sets of credentials, any of which can be a mix-and-match between written directly in `config.json`, specified as env vars or read from a file. E.g. the following would work:

```json
{
  "cloudflare": [
    {
      "authentication": {
          "api_token": "my_hard_coded_value", 
          "api_key": {
              "api_key": { "file": "/my/secret/file" },
              "account_email": { "env": "MY_ENV_VAR" }
          }
      },
      ...
    },
    {
      "authentication": {
          "api_token": { "file": "/my/other/secret/file" }, 
          "api_key": {
              "api_key": "my_other_hardcoded_value",
              "account_email": { "env": "MY_OTHER_ENV_VAR" }
          }
      },
      ...
    }
  ]
}
```

However, all of that does go against your preference of "I do not want to see changes to the current `config.json` format". Personally I like the idea of extending the format and keeping all of the config in one place, but it's not my project 😃 So, that leads to having environment variables only and ignoring values in `config.json` if the env vars are set (and in fact presumably the corresponding keys in `config.json` would be absent).

When you say having vars named `_2`, `_3`, ... `_10`, do you mean to correspond to the multiple domains that could be specified? If so I don't think we need to cap it at 10: for each config `i` we can just look for env vars ending with `_$i` (or +1 depending whether you want the env vars to be zero-indexed).

Have I understood your preference correctly above? (Or given you cause to reconsider extending the format? 😉)

**Comment by timothymiller** [comment]
Thanks for your work so far.

This PR needs to be synced with the latest master branch to be approved.

**Comment by dhrrgn** [comment]
I think this being over complicated, and overthought. 

**File Support**
In my opinion, file support is of little real world value. If you are already copying files, exporting those values to env vars is trivial (`export FOO=$(cat val.txt)`, `export $(cat .env | xargs)`, use `jq`, etc). So supporting env vars should make this unnecessary.

**Environment Variables**
Instead of predefined var names, with unhelpful indexes, I propose that it support any env var with a predefined prefix (e.g. `CF_DDNS_`). Then it's just a simple matter of using template substitution when reading config values that support env vars.

Example:

```python
import os
from collections import Collection, Mapping
from string import Template

# Read in all environment variables that have the correct prefix
ENV_VARS = {key: value for (key, value) in os.environ.items() if key.startswith('CF_DDNS_')}

# Stole this from StackOverflow for demo purposes
def recursive_map(data, func):
    apply = lambda x: recursive_map(x)
    if isinstance(data, Mapping):
        return type(data)({k: apply(v) for k, v in data.items()})
    elif isinstance(data, Collection):
        return type(data)(apply(v) for v in data)
    else:
        return func(data)

# Substitutes env variables in a template string with their values 
def resolve_env(val):
  return Template(val).safe_substitute(**ENV_VARS)

# Resolve env vars in all values
config = recursive_map(config, resolve_env)
```

This way, you can just do something like:

```
{
  "cloudflare": [
    {
      "authentication": {
          "api_token": "my_hard_coded_value", 
          "api_key": {
              "account_email": "$CF_DDNS_EMAIL"
          }
      },
      ...
  ]
}
```

**Comment by timothymiller** [comment]
> I think this being over complicated, and overthought.
> 
> **File Support** In my opinion, file support is of little real world value. If you are already copying files, exporting those values to env vars is trivial (`export FOO=$(cat val.txt)`, `export $(cat .env | xargs)`, use `jq`, etc). So supporting env vars should make this unnecessary.
> 
> **Environment Variables** Instead of predefined var names, with unhelpful indexes, I propose that it support any env var with a predefined prefix (e.g. `CF_DDNS_`). Then it's just a simple matter of using template substitution when reading config values that support env vars.
> 
> Example:
> 
> ```python
> import os
> from collections import Collection, Mapping
> from string import Template
> 
> # Read in all environment variables that have the correct prefix
> ENV_VARS = {key: value for (key, value) in os.environ.items() if key.startswith('CF_DDNS_')}
> 
> # Stole this from StackOverflow for demo purposes
> def recursive_map(data, func):
>     apply = lambda x: recursive_map(x)
>     if isinstance(data, Mapping):
>         return type(data)({k: apply(v) for k, v in data.items()})
>     elif isinstance(data, Collection):
>         return type(data)(apply(v) for v in data)
>     else:
>         return func(data)
> 
> # Substitutes env variables in a template string with their values 
> def resolve_env(val):
>   return Template(val).safe_substitute(**ENV_VARS)
> 
> # Resolve env vars in all values
> config = recursive_map(config, resolve_env)
> ```
> 
> This way, you can just do something like:
> 
> ```
> {
>   "cloudflare": [
>     {
>       "authentication": {
>           "api_token": "my_hard_coded_value", 
>           "api_key": {
>               "account_email": "$CF_DDNS_EMAIL"
>           }
>       },
>       ...
>   ]
> }
> ```

I like your suggestion. I opened [a call to action in the README](https://github.com/timothymiller/cloudflare-ddns#%EF%B8%8F-call-to-action-docker-environment-variable-support) for anyone interested in opening a PR to add this functionality.

**Comment by timothymiller** [comment]
This discussion is being carried on in this thread https://github.com/timothymiller/cloudflare-ddns/pull/117

**Pull request closed without merging**